File: 2024/day04.py
from pathlib import Path

def count_word_occurrences(word_search: list[list[str]], word: str):
    """`word_search` = list of rows"""
    word_length = len(word)
    n_occurrences = 0
    # Horizontal
    for row_no, _ in enumerate(word_search):
        for col_no in range(len(word_search[0]) - word_length + 1):
            s = [word_search[row_no][col_no + i] for i in range(word_length)]
            if ("".join(s) == word) or ("".join(reversed(s)) == word):
                n_occurrences += 1                
    # Vertical
    for row_no in range(0, len(word_search) - word_length + 1):
        for col_no, _ in enumerate(word_search[0]):
            s = [word_search[row_no + i][col_no] for i in range(word_length)]
            if ("".join(s) == word) or ("".join(reversed(s)) == word):
                n_occurrences += 1
    # Left-to-right diagonal
    for row_no in range(len(word_search) - word_length + 1):
        for col_no in range(len(word_search[0]) - word_length + 1):
            s = [word_search[row_no + i][col_no + i] for i in range(word_length)]
            if ("".join(s) == word) or ("".join(reversed(s)) == word):
                n_occurrences += 1
    # Right-to-left diagonal
    for row_no in range(len(word_search) - word_length + 1):    
        for col_no in range(word_length - 1, len(word_search[0])):
            s = [
                word_search[row_no + r][col_no + c]
                for r, c in zip(range(word_length), range(0, -word_length, -1))
            ]
            if ("".join(s) == word) or ("".join(reversed(s)) == word):
                n_occurrences += 1
    return n_occurrences


def count_word_crosses(word_search: list[list[str]], word: str):
    word_length = len(word)
    n_occurrences = 0
    for row_no in range(0, len(word_search) - word_length + 1):
        for col_no in range(0, len(word_search[0]) - word_length + 1):
            square = [[word_search[row_no + r][col_no + c] for c in range(0, word_length)] for r in range(0, word_length)]
            diag1 = [square[i][i] for i in range(0, word_length)]
            diag2 = [square[r][c] for r, c in zip(range(word_length), range(word_length - 1, -1, -1))]
            if (
                    ("".join(diag1) == word and "".join(diag2) == word)
                    or ("".join(reversed(diag1)) == word and "".join(diag2) == word)
                    or ("".join(diag1) == word and "".join(reversed(diag2)) == word)
                    or ("".join(reversed(diag1)) == word and "".join(reversed(diag2)) == word)
            ):
                n_occurrences += 1
    return n_occurrences


input_file = Path(__file__).parent / "input" / "day04.txt"
input_ = [list(l) for l in input_file.read_text().splitlines()]
print("Part 1:", count_word_occurrences(input_, "XMAS"))
print("Part 2:", count_word_crosses(input_, "MAS"))

File: 2024/src/day04.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};

type Row = Vec<char>;
type WordSearch = Vec<Row>;

fn reverse(s: String) -> String {
    s.chars()
        .rev()
        .map(|c| c.to_string())
        .collect::<Vec<String>>()
        .join("")
}

fn count_word_occurrences(word_search: WordSearch, word: String) -> i32 {
    let mut n_occurrences: i32 = 0;
    // Horizontal matches
    for row_no in 0..word_search.len() {
        for col_no in 0..(word_search[0].len() - word.len() + 1) {
            let v: String = (0..word.len())
                .map(|i| word_search[row_no][col_no + i].to_string())
                .collect::<Vec<_>>()
                .join("");

            if (v == word) || (reverse(v) == word) {
                n_occurrences += 1;
            }
        }
    }
    // Vertical matches
    for row_no in 0..(word_search.len() - word.len() + 1) {
        for col_no in 0..(word_search[0].len()) {
            let v: String = (0..word.len())
                .map(|i| word_search[row_no + i][col_no].to_string())
                .collect::<Vec<_>>()
                .join("");
            if (v == word) || (reverse(v) == word) {
                n_occurrences += 1;
            }
        }
    }
    // Left-to-right diagonal matches
    for row_no in 0..(word_search.len() - word.len() + 1) {
        for col_no in 0..(word_search[0].len() - word.len() + 1) {
            let v: String = (0..word.len())
                .map(|i| word_search[row_no + i][col_no + i].to_string())
                .collect::<Vec<_>>()
                .join("");
            if (v == word) || (reverse(v) == word) {
                n_occurrences += 1;
            }
        }
    }
    // Right-to-left diagonal matches
    for row_no in 0..(word_search.len() - word.len() + 1) {
        for col_no in (word.len() - 1)..word_search[0].len() {
            let v: String = (0..word.len())
                .map(|i| word_search[row_no + i][col_no - i].to_string())
                .collect::<Vec<_>>()
                .join("");
            if (v == word) || (reverse(v) == word) {
                n_occurrences += 1;
            }
        }
    }
    n_occurrences
}

fn count_word_crosses(word_search: WordSearch, word: String) -> i32 {
    let mut n_occurrences = 0;
    for row_no in 0..(word_search.len() - word.len() + 1) {
        for col_no in 0..(word_search[0].len() - word.len() + 1) {
            let mut square: WordSearch = Vec::new();
            for r in 0..word.len() {
                let mut row: Row = Vec::new();
                for c in 0..word.len() {
                    row.push(word_search[row_no + r][col_no + c])
                }
                square.push(row)
            }
            let diag1: String = (0..square.len())
                .map(|i| square[i][i].to_string())
                .collect::<Vec<String>>()
                .join("");
            let diag2: String = (0..square.len())
                .map(|i| square[i][word.len() - i - 1].to_string())
                .collect::<Vec<String>>()
                .join("");
            if ((diag1.clone() == word) && (diag2.clone() == word))
                || ((reverse(diag1.clone()) == word) && (diag2.clone() == word))
                || ((diag1.clone() == word) && (reverse(diag2.clone()) == word))
                || ((reverse(diag1.clone()) == word) && (reverse(diag2.clone()) == word))
            {
                n_occurrences += 1;
            }
        }
    }
    n_occurrences
}

pub fn main() {
    let input_path = env::current_dir().unwrap().join("input").join("day04.txt");
    let file = File::open(input_path);
    let reader = BufReader::new(file.unwrap());
    let input_: WordSearch = reader
        .lines()
        .map(|l| l.unwrap().chars().collect::<Row>())
        .collect();
    println!("{}", "Day 4".to_owned());
    println!(
        "Part 1: {}",
        count_word_occurrences(input_.clone(), "XMAS".to_owned())
    );
    println!(
        "Part 2: {}",
        count_word_crosses(input_.clone(), "MAS".to_owned())
    );
}
